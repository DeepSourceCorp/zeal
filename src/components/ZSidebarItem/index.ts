import ZSidebarItem from '../ZSidebarMenu/ZSidebarItem.vue'

export default ZSidebarItem
