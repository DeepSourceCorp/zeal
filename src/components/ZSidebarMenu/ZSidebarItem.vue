<template>
  <div class="sidebar-item flex space-x-2 px-2 py-2 items-center cursor-pointer hover:bg-ink-300 text-sm">
    <z-icon v-if="icon" 
            :icon="icon" 
            size="small"
            :color="iconColor"></z-icon>
    <div class="items-center space-x-1 w-full hidden sm:flex"><slot></slot></div>
  </div>
</template>

<script>
import ZIcon from '@/components/ZIcon/ZIcon.vue'
export default {
  name: 'ZSidebarMenuItem',
  components: {
    ZIcon
  },
  props: {
    icon: {
      type: String,
    },
    iconColor: {
      type: String
    }
  }
}
</script>
