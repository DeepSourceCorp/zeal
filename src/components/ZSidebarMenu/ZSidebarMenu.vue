<template>
  <div>
    <div
      v-outside-click="closeModal"
      v-on:click="openModal()"
      class="absolute z-10 block cursor-pointer top-2 left-2 lg:hidden"
      :class="{ hidden: isOpen }"
    >
      <z-icon icon="menu"></z-icon>
    </div>
    <div
      class="absolute top-0 z-30 flex flex-col h-screen transition-all duration-300 ease-in-out cursor-pointer group bg-ink-400 sidebar-menu border-ink-200 text-vanilla-100"
      :class="[getWidth, getDirectionClasses, getBorderClasses, isOpen && 'shadow-black']"
    >
      <header
        class="w-full p-2 border-b border-solid border-ink-200"
        v-if="$scopedSlots.header"
        :class="[isCollapsed ? 'h-20' : 'h-12']"
      >
        <slot name="header" :isCollapsed="isCollapsed"></slot>
      </header>
      <div class="flex-1 w-full p-2 overflow-scroll sidebar-items">
        <slot :isCollapsed="isCollapsed"></slot>
      </div>
      <footer
        class="w-full px-2 py-4 border-t border-solid border-ink-200"
        v-if="$scopedSlots.footer"
      >
        <slot name="footer" :isCollapsed="isCollapsed"></slot>
      </footer>
      <div
        class="absolute top-0 hidden w-px h-full lg:group-hover:block -right-px bg-gradient-juniper_gradient"
      ></div>
      <div
        class="lg:group-hover:block hidden bg-juniper p-0.5 rounded-full relative lg:absolute top-3"
        @click.stop="collapseSidebar()"
        :class="getHoverStyle"
      >
        <z-icon :icon="getArrow" size="small" color="ink-400"></z-icon>
      </div>
    </div>
    <!-- Overlay -->
    <div
      :class="{ 'absolute w-full h-screen bg-ink-400 opacity-50 left-0 top-0 z-20': isOpen }"
    ></div>
  </div>
</template>

<script>
import Vue from 'vue'
import ZIcon from '@/components/ZIcon/ZIcon.vue'
import outsideClickDirective from '@/directives/outside-click'

export default {
  name: 'ZSidebarMenu',
  components: {
    ZIcon
  },
  directives: {
    outsideClick: outsideClickDirective
  },
  props: {
    width: {
      type: String,
      default: '64'
    },
    collapsed: {
      type: Boolean,
      default: false
    },
    widthCollapsed: {
      type: String,
      default: '12'
    },
    direction: {
      type: String,
      default: 'left'
    }
  },
  mounted() {
    this.$nextTick(() => {
      window.addEventListener('resize', () => {
        if (this.windowScreenWidth < this.largeScreenSize) this.isOpen = false
      })
    })
  },
  data() {
    return {
      isCollapsed: this.collapsed,
      isOpen: false,
      largeScreenSize: 1024,
      windowScreenWidth: window.innerWidth
    }
  },
  computed: {
    sidebarWidth() {
      return this.isCollapsed ? this.widthCollapsed : this.width
    },
    getDirectionClasses() {
      const directionStyle = {
        left: 'lg:left-0 -left-3/4',
        right: 'lg:right-0 -right-3/4'
      }
      const openDirections = {
        left: 'left-0',
        right: 'right-0'
      }
      if (this.isOpen) {
        return openDirections[this.direction]
      }
      return directionStyle[this.direction]
    },
    getBorderClasses() {
      const borders = {
        left: 'border-r',
        right: 'border-l'
      }
      return !this.isOpen && borders[this.direction]
    },
    getWidth() {
      if (this.isCollapsed) return `lg:w-${this.widthCollapsed} w-9/12`
      return `lg:w-${this.width} w-9/12`
    },
    getHoverStyle() {
      const directionStyle = {
        left: '-right-3',
        right: '-left-3'
      }
      return directionStyle[this.direction]
    },
    getArrow() {
      if (
        (this.direction == 'left' && !this.isCollapsed) ||
        (this.direction == 'right' && this.isCollapsed)
      ) {
        return `chevron-left`
      }
      return `chevron-right`
    }
  },
  methods: {
    collapseSidebar() {
      this.isCollapsed = !this.isCollapsed
    },
    openModal() {
      this.isOpen = !this.isOpen
    },
    closeModal() {
      this.isOpen = false
    }
  }
}
</script>

