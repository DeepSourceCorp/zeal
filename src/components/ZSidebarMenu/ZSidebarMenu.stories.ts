import '@/assets/css/layout.css'
import '@/assets/css/typography.css'
import '@/assets/css/tailwind.css'
import ZSidebarMenu from './ZSidebarMenu.vue'
import ZSidebarItem from '@/components/ZSidebarMenu/ZSidebarItem.vue'
import ZAvatar from '@/components/ZAvatar/ZAvatar.vue'
import ZIcon from '@/components/ZIcon/ZIcon.vue'
import ZButton from '@/components/ZButton/ZButton.vue'

export default {
  title: 'Sidebar Menu',
  component: ZSidebarMenu,
  excludeStories: /.*Data$/
}

export const DefaultSidebarMenu = () => ({
  components: { ZSidebarMenu, ZSidebarItem, ZIcon, ZAvatar, ZButton },
  template: `
  <div class="w-80">  
    <z-sidebar-menu>
      <div slot="header" class="h-full flex flex-col sm:flex-row items-center p-1 justify-between">
        <div class="flex space-x-1 items-center">
          <z-icon icon="globe" size="small"></z-icon>
          <span class="hidden sm:block">MediaCorp</span>
        </div>
        <div class="flex sm:space-x-1 items-center">
          <z-avatar
            image="https://randomuser.me/api/portraits/women/24.jpg"
            user-name="Akshay Paliwal"
            size="sm"
            class="hidden sm:block"
          ></z-avatar>
          <z-avatar
            image="https://randomuser.me/api/portraits/women/24.jpg"
            user-name="Akshay Paliwal"
            size="xs"
            class="block sm:hidden"
          ></z-avatar>
          <z-icon icon="chevron-down" size="small" class="hidden sm:block"></z-icon>
        </div>
      </div>
      <z-sidebar-item icon="columns">My dashboard</z-sidebar-item>
      <z-sidebar-item icon="home">Team home</z-sidebar-item>
      <z-sidebar-item icon="activity">
        <span class="flex-1">Actively Analyzing</span>
        <z-button color="primary" 
            icon="plus" 
            iconColor="ink-400" 
            iconSize="small">
        </z-button>
      </z-sidebar-item>
      <z-sidebar-item icon="archive">All repositories</z-sidebar-item>
      <div slot="footer" class="h-full flex flex-col justify-evenly">
        <z-sidebar-item icon="users">My team</z-sidebar-item>
        <z-sidebar-item icon="settings">Team settings</z-sidebar-item>
        <z-sidebar-item icon="star" icon-color="juniper">Upgrade to Pro</z-sidebar-item>
      </div>
    </z-sidebar-menu>
  </div>`
})
