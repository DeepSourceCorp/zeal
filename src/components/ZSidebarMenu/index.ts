import ZSidebarMenu from './ZSidebarMenu.vue'

export default ZSidebarMenu
