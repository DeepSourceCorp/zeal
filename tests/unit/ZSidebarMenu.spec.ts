import ZSidebarMenu from '../../src/components/ZSidebarMenu'
import ZSidebarItem from '../../src/components/ZSidebarItem'
import { mount } from '@vue/test-utils'

const defaultSidebar = {
  template: `
        <z-sidebar-menu>
        <div slot="header">
            <span>Deepsource</span>
        </div>
        <z-sidebar-item icon="globe">asgard</z-sidebar-item>
        <z-sidebar-item icon="git-merge">asgard-office</z-sidebar-item>
        <z-sidebar-item icon="git-merge">second-release</z-sidebar-item>
        <z-sidebar-item icon="git-merge">another-release</z-sidebar-item>
        <div slot="footer">
            <z-sidebar-item icon="users">My team</z-sidebar-item>
            <z-sidebar-item icon="settings">Team settings</z-sidebar-item> 
        </div>
        </z-sidebar-menu>`,
  components: {
    ZSidebarMenu,
    ZSidebarItem
  }
}

describe('ZMenu', () => {
  it('renders a sidebar in the specified direction', async () => {
    const wrapper = mount(defaultSidebar)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
