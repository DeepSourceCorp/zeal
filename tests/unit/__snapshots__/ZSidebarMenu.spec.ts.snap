// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZMenu renders a sidebar in the specified direction 1`] = `
<div>
  <div class="absolute top-2 left-2 block lg:hidden cursor-pointer z-10" data-outside-click-index="0"><svg viewBox="0 0 24 24" fill="none" class="stroke-2 z-icon w-4 h-4 z-icon--menu text-vanilla-400">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg></div>
  <div class="group bg-ink-400 absolute z-20 sidebar-menu h-screen flex flex-col border-ink-200 text-vanilla-100 top-0 cursor-pointer transition-all duration-300 ease-in-out lg:w-64 w-9/12 lg:left-0 -left-3/4 border-r">
    <header class="w-full border-b border-solid border-ink-200 p-2 h-12">
      <div><span>Deepsource</span></div>
    </header>
    <div class="sidebar-items w-full flex-1 p-2 overflow-scroll">
      <div class="sidebar-item flex space-x-2 px-2 py-2 items-center cursor-pointer hover:bg-ink-300 text-sm transition-all duration-300 ease-in-out"><svg viewBox="0 0 24 24" fill="none" class="stroke-2 z-icon w-4 h-4 z-icon--globe text-vanilla-400">
          <circle cx="12" cy="12" r="10"></circle>
          <line x1="2" y1="12" x2="22" y2="12"></line>
          <path d="M12 2a15.3 15.3 0 0 1 4 10 15.3 15.3 0 0 1-4 10 15.3 15.3 0 0 1-4-10 15.3 15.3 0 0 1 4-10z"></path>
        </svg>
        <div class="items-center space-x-1 w-full text-xs sm:text-sm flex">asgard</div>
      </div>
      <div class="sidebar-item flex space-x-2 px-2 py-2 items-center cursor-pointer hover:bg-ink-300 text-sm transition-all duration-300 ease-in-out"><svg viewBox="0 0 24 24" fill="none" class="stroke-2 z-icon w-4 h-4 z-icon--git-merge text-vanilla-400">
          <circle cx="18" cy="18" r="3"></circle>
          <circle cx="6" cy="6" r="3"></circle>
          <path d="M6 21V9a9 9 0 0 0 9 9"></path>
        </svg>
        <div class="items-center space-x-1 w-full text-xs sm:text-sm flex">asgard-office</div>
      </div>
      <div class="sidebar-item flex space-x-2 px-2 py-2 items-center cursor-pointer hover:bg-ink-300 text-sm transition-all duration-300 ease-in-out"><svg viewBox="0 0 24 24" fill="none" class="stroke-2 z-icon w-4 h-4 z-icon--git-merge text-vanilla-400">
          <circle cx="18" cy="18" r="3"></circle>
          <circle cx="6" cy="6" r="3"></circle>
          <path d="M6 21V9a9 9 0 0 0 9 9"></path>
        </svg>
        <div class="items-center space-x-1 w-full text-xs sm:text-sm flex">second-release</div>
      </div>
      <div class="sidebar-item flex space-x-2 px-2 py-2 items-center cursor-pointer hover:bg-ink-300 text-sm transition-all duration-300 ease-in-out"><svg viewBox="0 0 24 24" fill="none" class="stroke-2 z-icon w-4 h-4 z-icon--git-merge text-vanilla-400">
          <circle cx="18" cy="18" r="3"></circle>
          <circle cx="6" cy="6" r="3"></circle>
          <path d="M6 21V9a9 9 0 0 0 9 9"></path>
        </svg>
        <div class="items-center space-x-1 w-full text-xs sm:text-sm flex">another-release</div>
      </div>
    </div>
    <footer class="w-full border-t border-solid border-ink-200 px-2 py-4">
      <div>
        <div class="sidebar-item flex space-x-2 px-2 py-2 items-center cursor-pointer hover:bg-ink-300 text-sm transition-all duration-300 ease-in-out"><svg viewBox="0 0 24 24" fill="none" class="stroke-2 z-icon w-4 h-4 z-icon--users text-vanilla-400">
            <path d="M17 21v-2a4 4 0 0 0-4-4H5a4 4 0 0 0-4 4v2"></path>
            <circle cx="9" cy="7" r="4"></circle>
            <path d="M23 21v-2a4 4 0 0 0-3-3.87"></path>
            <path d="M16 3.13a4 4 0 0 1 0 7.75"></path>
          </svg>
          <div class="items-center space-x-1 w-full text-xs sm:text-sm flex">My team</div>
        </div>
        <div class="sidebar-item flex space-x-2 px-2 py-2 items-center cursor-pointer hover:bg-ink-300 text-sm transition-all duration-300 ease-in-out"><svg viewBox="0 0 24 24" fill="none" class="stroke-2 z-icon w-4 h-4 z-icon--settings text-vanilla-400">
            <circle cx="12" cy="12" r="3"></circle>
            <path d="M19.4 15a1.65 1.65 0 0 0 .33 1.82l.06.06a2 2 0 0 1 0 2.83 2 2 0 0 1-2.83 0l-.06-.06a1.65 1.65 0 0 0-1.82-.33 1.65 1.65 0 0 0-1 1.51V21a2 2 0 0 1-2 2 2 2 0 0 1-2-2v-.09A1.65 1.65 0 0 0 9 19.4a1.65 1.65 0 0 0-1.82.33l-.06.06a2 2 0 0 1-2.83 0 2 2 0 0 1 0-2.83l.06-.06a1.65 1.65 0 0 0 .33-1.82 1.65 1.65 0 0 0-1.51-1H3a2 2 0 0 1-2-2 2 2 0 0 1 2-2h.09A1.65 1.65 0 0 0 4.6 9a1.65 1.65 0 0 0-.33-1.82l-.06-.06a2 2 0 0 1 0-2.83 2 2 0 0 1 2.83 0l.06.06a1.65 1.65 0 0 0 1.82.33H9a1.65 1.65 0 0 0 1-1.51V3a2 2 0 0 1 2-2 2 2 0 0 1 2 2v.09a1.65 1.65 0 0 0 1 1.51 1.65 1.65 0 0 0 1.82-.33l.06-.06a2 2 0 0 1 2.83 0 2 2 0 0 1 0 2.83l-.06.06a1.65 1.65 0 0 0-.33 1.82V9a1.65 1.65 0 0 0 1.51 1H21a2 2 0 0 1 2 2 2 2 0 0 1-2 2h-.09a1.65 1.65 0 0 0-1.51 1z"></path>
          </svg>
          <div class="items-center space-x-1 w-full text-xs sm:text-sm flex">Team settings</div>
        </div>
      </div>
    </footer>
    <div class="hidden lg:group-hover:block absolute top-0 right-0 h-full w-px bg-gradient-to-t from-vanilla-300 via-juniper to-juniper"></div>
    <div class="lg:group-hover:block hidden bg-juniper p-1 rounded-full relative lg:absolute top-3 -right-3"><svg viewBox="0 0 24 24" fill="none" class="stroke-2 z-icon w-4 h-4 z-icon--chevron-left text-ink-400">
        <polyline points="15 18 9 12 15 6"></polyline>
      </svg></div>
  </div>
  <div class=""></div>
</div>
`;
